lab/skel_graph/shortest_path/dijkstra/main.cpp

```diff
@@ -1,157 +1,179 @@
-// Copyright 2018 Darius Neatu <neatudarius@gmail.com>
-
-
-/*
-    skel_graph - Dijkstra - O(m log n)
-    (Sau cu notatia din lab O(E log V))
-    http://www.infoarena.ro/problema/dijkstra
-
-    Deoarece folosesc heapuri binare (priority_queue),
-    complexitatea este cea de mai sus
-*/
-
-#include <algorithm>
-#include <bitset>
-#include <cassert>
-#include <cstdio>
-#include <cstring>
-#include <fstream>
-#include <iostream>
-#include <queue> // std::priority_queue
-#include <vector>
-#define NMAX 50009   // ATENTIE la cat e in problema curenta !!!
-#define MMAX 250009  // ATENTIE la cat e in problema curenta !!!
+// Copyright 2018 Popescu Alexandru Gabriel <alexandru.popescu0697@gmail.com>
+// Copyright 2017 Darius Neatu <neatudarius@gmail.com>
+
+// Dijkstra
+// O(m log n)
+// Deoarece folosesc heapuri binare (priority_queue),
+// complexitatea este cea de mai sus
+
+// http://www.infoarena.ro/problema/dijkstra
+
+#include <bits/stdc++.h>
+
+#define NMAX 50010   // ATENTIE la cat e in problema curenta !!!
 #define oo (1 << 30) // ATENTIE la cat e in problema curenta !!!
+#define NO_PARENT (-1)
+
 using namespace std;
 
-// tipul edge
-// muchia de cost "cost", de la node la neighbour,
-// va fi un element edge(neighbour, cost) in G[node]
-typedef pair<int, int> edge;
-#define neighbour first
-#define cost second
-
-// aceleasi semnificatii ca la BFS
-int n, m;
-int d[NMAX], p[NMAX];
-vector<edge> G[NMAX];
-
-// declararea unui priority queue in care pun numere intregi (noduri)
-// si cu un criteriu custom de prioritate
-
-// varianta 1
-// http://www.cplusplus.com/reference/queue/priority_queue/priority_queue/
-struct cmp {
-  // returneaza true daca x ar trebui "sa fie sub" y in arbore
-  bool operator()(const int &x, const int &y) const { return d[x] > d[y]; }
-};
-priority_queue<int, vector<int>, cmp> pq;
+class Task {
+public:
+  void solve() {
+    read_input();
+    get_result();
+    print_output();
+  }
 
-// varianta 2 - C++11 :D
-// http://en.cppreference.com/w/cpp/language/decltype
-// bool cmp(const int &x, const int &y) {
-//     // returneaza true daca x ar trebui "sa fie sub" y in arbore
-//     return d[x] > d[y];
-// }
-// priority_queue<int, vector<int>, decltype(&cmp)> pq(cmp);
+private:
+  // n = numar de noduri, m = numar de muchii
+  int n, m;
 
-void read_input() {
-  cin >> n >> m;
+  int source;
 
-  for (int i = 1; i <= m; ++i) {
-    int x, y, c; // x -> y, de cost c
-    cin >> x >> y >> c;
+  // adj[i] contine perechi de tipul : vecin , costul muchiei i -> vecin
+  vector<pair<int, int>> adj[NMAX];
 
-    G[x].push_back(edge(y, c));
-    // G[y].push_back(edge(x, c)); // sterg daca e orientat !!!
-  }
-}
+  // say something
+  priority_queue<pair<int, int>, vector<pair<int, int>>,
```

Comment:
```
#define edge pair<int, int>
```
I think this will make the code more readable.
or better
```
priority_queue<pair<int, int>> pq(greater<pair<int, int>>());
```
Is this more readable?

Reply:
The performance should be the same!

Reply:
Did you try this?

Reply:
Did you try this?